### PR TITLE
Make build_web_compilers output full kernel in debug mode

### PR DESCRIPTION
**Pull request by annagrin**
Make build_web_compilers output full kernel files and fixup metadata paths, in preparation for expression 
evaluation feature in webdev:

- make build_web_compilers output full kernel files in debug mode
  This is preparation work for expression evaluation in webdev.
  depends on the following SDK change:
   https://dart-review.googlesource.com/c/sdk/+/160944

- update paths inside metadata to match buildstep copying them to a final
  location

### CODE REVIEW

File: build_web_compilers/lib/src/dev_compiler_builder.dart
Comment (jakemac53) [inline build_web_compilers/lib/src/dev_compiler_builder.dart:235]: nit: unnecessary comment

File: build_web_compilers/lib/src/dev_compiler_builder.dart
Comment (jakemac53) [inline build_web_compilers/lib/src/dev_compiler_builder.dart:250]: nit: I would be a bit more clear here:

```dart
// Copy the metadata output, modifying its contents to remove the temp directory from paths
```

Also, I would expect to see the `multi-root` scheme or something here, the metadata files might need to be updated - this will be harder to clean up in blaze I think (there isn't just one possible dir to strip out).

File: build_web_compilers/lib/src/dev_compiler_builder.dart
Comment (jakemac53) [inline build_web_compilers/lib/src/dev_compiler_builder.dart:316]: It doesn't look like the `path` arg is ever passed anything but `''` here, might as well remove it.

Also i would make this a private method.

**Comment by natebosch** [comment]
> We should probably be hiding this behavior behind some special flag, at least for now.

The downside to a flag is that we'll have to remove it when we want this to be the default.

File: build_web_compilers/lib/src/dev_compiler_builder.dart
Comment (annagrin) [reply build_web_compilers/lib/src/dev_compiler_builder.dart:250]: Is the same code running in blaze?

File: build_web_compilers/lib/src/dev_compiler_builder.dart
Comment (annagrin) [reply build_web_compilers/lib/src/dev_compiler_builder.dart:316]: will do, thanks!

File: build_web_compilers/lib/src/dev_compiler_builder.dart
Comment (natebosch) [reply build_web_compilers/lib/src/dev_compiler_builder.dart:250]: No `build_web_compilers` is external only. The blaze integration uses custom starlark build rules, it doesn't go through the `Builder` interface.

**Comment by annagrin** [comment]
> You'll need to update the min SDK and update the expects in tests as they will now fail with the new files.

Still working on my SDK changes, will update as soon as the PR lands.

File: build_web_compilers/lib/src/dev_compiler_builder.dart
Comment (jakemac53) [reply build_web_compilers/lib/src/dev_compiler_builder.dart:250]: Right, I just mean that if you are having to write this here it will be harder to replicate there.

**Comment by jakemac53** [comment]
> The downside to a flag is that we'll have to remove it when we want this to be the default.

I am not sure this should ever be the default though - we potentially only want to enable it conditionally? Or at least have the option to disable I think.

**Comment by natebosch** [comment]
> I am not sure this should ever be the default though - we potentially only want to enable it conditionally? Or at least have the option to disable I think.

I'm not sure what that would look like or when we'd tell people to use it? Do you imagine that if folks complain about memory use or performance we'd tell them to pass an option but if they do they lose expression evaluation?

I'd think by default we would want expression evaluation to work.

**Comment by grouma** [comment]
> I'm not sure what that would look like or when we'd tell people to use it? Do you imagine that if folks complain about memory use or performance we'd tell them to pass an option but if they do they lose expression evaluation?

Do we anticipate this having much of an impact on performance? My understanding is that DDC already computes this information already and it is simply dumping the information to an additional file. Shouldn't that have minimal impact?

**Comment by jakemac53** [comment]
> Do we anticipate this having much of an impact on performance? My understanding is that DDC already computes this information already and it is simply dumping the information to an additional file. Shouldn't that have minimal impact?

We end up serializing it to a file once in DDC, and then copying it back inside the Builder, so we write the file at least twice. Then there is the complication that we also process the file to md5 hash it, and we cache it in memory.

File: build_web_compilers/lib/src/dev_compiler_builder.dart
Comment (annagrin) [reply build_web_compilers/lib/src/dev_compiler_builder.dart:250]: I need to change the ddc code to generate metadata with a multiroot scheme, currently it has full paths inside, which is not what we what for part of build output. 

Created an issue to modify metadata: https://github.com/dart-lang/sdk/issues/43309
Created an issue to update the code according to the change in build_web_compilers: https://github.com/dart-lang/build/issues/2802

**Comment by annagrin** [comment]
@natebosch re:flag: 

We can follow the same strategy we did in flutter - hide the functionality under a flag, at first disabled by default. Experiment, dogfood, and fix until the feature is stable enough.  

After that, make the flag enabled by default  but do not remove the flag so the users have the option to disable the feature if causes problems. We can eventually remove the flag when the feature becomes super stable, if we prefer to have less flags in the system.

Does this sound reasonable?

File: build_web_compilers/lib/src/dev_compiler_builder.dart
Comment (grouma) [inline build_web_compilers/lib/src/dev_compiler_builder.dart:37]: This package isn't used inside Google3 at all so we can probably leave out this extra bit of detail.

**Review by grouma — APPROVED** [review]
LGTM. Let's wait for approval from @natebosch or @jakemac53 first though.

File: build_web_compilers/lib/src/dev_compiler_builder.dart
Comment (natebosch) [inline build_web_compilers/lib/src/dev_compiler_builder.dart:30]: [nit] Should use a noun phrase.

```suggestion
  /// Whether to generate full dill file outputs for each module.
```

File: build_web_compilers/lib/src/dev_compiler_builder.dart
Comment (natebosch) [inline build_web_compilers/lib/src/dev_compiler_builder.dart:32]: This comment about "debug mode" feels misplaced. At best we can say "Full dill is only generated by default in debug mode," but I don't think even that is true, since we haven't updated the `build.yaml` file to set this option by default.

File: build_web_compilers/lib/src/dev_compiler_builder.dart
Comment (natebosch) [inline build_web_compilers/lib/src/dev_compiler_builder.dart:35]: I think somewhere in this comment we should mention that it implies an extra file. As currently phrased it feels to me like different content in the same file.

File: build_web_compilers/lib/src/dev_compiler_builder.dart
Comment (annagrin) [reply build_web_compilers/lib/src/dev_compiler_builder.dart:32]: the comment is no longer true, I will remove it

Comment (annagrin) [reply build_web_compilers/lib/src/dev_compiler_builder.dart:32]: I got confused by debug mode - apparently debug mode settings is always on in my case since ddc is only used in debug mode. I will remove the comment.

File: build_web_compilers/lib/src/dev_compiler_builder.dart
Comment (annagrin) [reply build_web_compilers/lib/src/dev_compiler_builder.dart:35]: will do

File: build_web_compilers/lib/src/dev_compiler_builder.dart
Comment (annagrin) [reply build_web_compilers/lib/src/dev_compiler_builder.dart]: thanks, fixed

File: build_web_compilers/lib/src/dev_compiler_builder.dart
Comment (annagrin) [reply build_web_compilers/lib/src/dev_compiler_builder.dart]: thanks, fixed

File: build_web_compilers/lib/src/dev_compiler_builder.dart
Comment (annagrin) [reply build_web_compilers/lib/src/dev_compiler_builder.dart]: fixed

**Review by jakemac53 — APPROVED** [review]

**Review by natebosch — APPROVED** [review]

**Comment by annagrin** [comment]
Thanks everyone!

**Pull request merged**
